workflow/gccontroller/gc_controller.go

```diff
@@ -82,6 +156,43 @@ func (c *Controller) runWorker() {
 	}
 }
 
+// retentionGC queues workflows for deletion based upon the retention policy.
+func (c *Controller) runCompletedGC() {
+	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
+	c.completedOrderedQueue.Lock()
+	for c.completedOrderedQueue.Len() > c.retentionPolicy.Completed {
+		key, _ := cache.MetaNamespaceKeyFunc(heap.Pop(c.completedOrderedQueue))
+		log.Infof("Queueing %v workflow %s for delete due to max rention(%d workflows)", wfv1.WorkflowSucceeded, key, c.retentionPolicy.Completed)
+		c.workqueue.Add(key)
+		<-c.limiter
+	}
+	c.completedOrderedQueue.Unlock()
+}
+
+func (c *Controller) runFailedGC() {
+	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
+	c.failedOrderedQueue.Lock()
+	for c.failedOrderedQueue.Len() > c.retentionPolicy.Failed {
+		key, _ := cache.MetaNamespaceKeyFunc(heap.Pop(c.failedOrderedQueue))
+		log.Infof("Queueing %v workflow %s for delete due to max rention(%d workflows)", wfv1.WorkflowFailed, key, c.retentionPolicy.Failed)
+		c.workqueue.Add(key)
+		<-c.limiter
+	}
+	c.failedOrderedQueue.Unlock()
+}
+
+func (c *Controller) runErrorGC() {
+	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
```

Comment:
these 3 funcs dupe much code, by using a map for the queues, and a single lock, you can have a single func, you can then use one limiter, and that limiter could be a var, rather than a struct field